File: app/src/main/java/de/westnordost/osmagent/data/osm/persist/AOsmElementDao.java
package de.westnordost.osmagent.data.osm.persist;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import de.westnordost.osmapi.map.data.Element;

public abstract class AOsmElementDao<T extends Element>
{
	private final SQLiteOpenHelper dbHelper;

	public AOsmElementDao(SQLiteOpenHelper dbHelper)
	{
		this.dbHelper = dbHelper;
	}

	/* Adds or updates the given object to the database */
	public void put(T object)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		db.insertWithOnConflict(getTableName(), null, createContentValuesFrom(object),
				SQLiteDatabase.CONFLICT_REPLACE);
	}

	public void delete(long id)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();
		db.delete(getTableName(), getIdColumnName() + " = " + id, null);
	}

	public T get(long id)
	{
		SQLiteDatabase db = dbHelper.getReadableDatabase();
		Cursor cursor = db.query(getTableName(), null, getIdColumnName() + " = " + id,
					null, null, null, null, "1");
		try
		{
			if (!cursor.moveToFirst()) return null;
			return createObjectFrom(cursor);
		}
		finally
		{
			cursor.close();
		}
	}

	/** Cleans up element entries that are not referenced by any quest anymore. */
	public void deleteUnreferenced()
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();

		String where = NodeTable.Columns.ID + " NOT IN ( " +
				"SELECT " + OsmQuestTable.Columns.ELEMENT_ID + " AS " + getIdColumnName() + " " +
				"FROM " + OsmQuestTable.NAME + " " +
				"WHERE " + OsmQuestTable.Columns.ELEMENT_TYPE + " = \"" + getElementTypeName() +"\"" +
				")";

		db.delete(getTableName(), where, null);
	}

	protected abstract String getElementTypeName();

	protected abstract String getTableName();
	protected abstract String getIdColumnName();

	protected abstract ContentValues createContentValuesFrom(T object);
	protected abstract T createObjectFrom(Cursor cursor);
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/AOsmElementDao.kt
package de.westnordost.streetcomplete.data.osm.mapdata

import android.database.sqlite.SQLiteOpenHelper

import de.westnordost.osmapi.map.data.Element
import de.westnordost.streetcomplete.data.ObjectRelationalMapping
import de.westnordost.streetcomplete.data.osm.changes.OsmElementChangesTable
import de.westnordost.streetcomplete.ktx.query
import de.westnordost.streetcomplete.ktx.queryOne
import de.westnordost.streetcomplete.ktx.transaction

/** Abstract base class for the DAOs that store the OSM elements */
abstract class AOsmElementDao<T : Element>(private val dbHelper: SQLiteOpenHelper) {

    private val db get() = dbHelper.writableDatabase

    protected abstract val elementTypeName: String
    protected abstract val tableName: String
    protected abstract val idColumnName: String
    protected abstract val lastUpdateColumnName: String
    protected abstract val mapping: ObjectRelationalMapping<T>

    fun put(element: T) {
        db.replaceOrThrow(tableName, null, mapping.toContentValues(element))
    }

    fun get(id: Long): T? {
        return db.queryOne(tableName, null, "$idColumnName = $id", null) { mapping.toObject(it) }
    }

    fun delete(id: Long) {
        db.delete(tableName, "$idColumnName = $id", null)
    }

    fun putAll(elements: Collection<T>) {
        if (elements.isEmpty()) return
        db.transaction {
            for (element in elements) {
                put(element)
            }
        }
    }

    fun getAll(ids: Collection<Long>): List<T> {
        if (ids.isEmpty()) return emptyList()
        return db.query(tableName, null, "$idColumnName IN (${ids.joinToString(",")})") { mapping.toObject(it) }
    }

    fun deleteAll(ids: Collection<Long>) {
        if (ids.isEmpty()) return
        db.delete(tableName, "$idColumnName IN (${ids.joinToString(",")})", null)
    }

    /** Get all elements that are not referenced by any quest (or other things) anymore and older
     *  than the given timestamp*/
    open fun getUnusedAndOldIds(timestamp: Long): List<Long> {
        return db.query(tableName, arrayOf(idColumnName), """
            $lastUpdateColumnName < $timestamp AND
            $idColumnName NOT IN (
                SELECT ${OsmElementChangesTable.Columns.ELEMENT_ID} AS $idColumnName
                FROM ${OsmElementChangesTable.NAME}
                WHERE ${OsmElementChangesTable.Columns.ELEMENT_TYPE} = "$elementTypeName"
            )""".trimIndent()) {
            it.getLong(0)
        }
    }
}


